app/model/PaymentSpecificData.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package model
+import play.api.libs.json._
+
+trait PaymentSpecificData {
+  def getReference: String
+}
+
+case class PaymentSpecificDataP800(
+    referencePart1: String,
+    referencePart2: String,
+    referencePart3: String,
+    period:         Int
+) extends PaymentSpecificData {
+  def getReference: String = {
+    s"$referencePart1$referencePart2$referencePart3"
+  }
+}
+
+object PaymentSpecificDataP800 {
+  implicit val format: OFormat[PaymentSpecificDataP800] = Json.format[PaymentSpecificDataP800]
+
+}
+
+object PaymentSpecificData {
+  implicit val writes: Writes[PaymentSpecificData] = Writes[PaymentSpecificData] {
+    case psd: PaymentSpecificDataP800 => PaymentSpecificDataP800.format.writes(psd)
+    case _                            => throw new RuntimeException("Unsupported write")
+  }
+
+  implicit val read: Reads[PaymentSpecificData] = Reads[PaymentSpecificData] {
+    case psd: PaymentSpecificDataP800 => PaymentSpecificDataP800.format.reads(psd)
```

Comment:
I'm afraid this won't work.

Reads is a "method" of reading json into something. There must be in json field which designates what is the "type" of this json and what other fields to expect so all json can be mapped to specific class.

I'll post in another comment example code for that.

Reply:
I don't think the reads is needed .

Reply:
No code posted but I there is an example of how to do this in pay-frontend. 
See the `Journey[+Jsd <: JourneySpecificData]` (which is equivalent of your `TpsPaymentItem`)
And see the `JourneySpecificData` (which is equivalent of your `PaymentSpecificData`)

Reply:
hmmm, still reading rest of this PR, maybe above comment is not needed.

Reply:
Cool, let's remove this dead code then.